Added support for enum.protocols feature

aaronsky opened the pull request:
Addresses #171.

Creates a new `enum.protocols` option that is used to drive protocol conformance for enums of strings, which are the only currently supported enum variety. Currently does not do any of the redundant protocol stripping that `entities.protocols` does.

aaronsky committed:
Added support for enum.protocols feature

liamnichols commented on Sources/CreateOptions/ConfigOptions.swift:
```diff
@@ -568,6 +568,18 @@ public struct ConfigOptions: ParsableConfiguration {
+    @Option public var enums: Enums
+
+    // sourcery: document,
+    /// Options specifically related to generating enums of strings.
```
Just a thought, but #184 might mean that we generated enums from integer properties too so maybe worth rewording this:

```suggestion
    /// Options specifically related to enums generated from schemas.
```

liamnichols commented on Sources/CreateOptions/ConfigOptions.swift:
```diff
@@ -568,6 +568,18 @@ public struct ConfigOptions: ParsableConfiguration {
+    // sourcery: document,
+    /// Options specifically related to generating enums of strings.
+    ///
+    /// Note: these options do not apply to other enums, such as `CodingKey` conformances or
```
Just a suggestion to get the nicer formatting in GitHub

```suggestion
    /// > **Note**: these options do not apply to other enums, such as `CodingKey` conformances or
```

It would look like this:

> **Note**: these options do not apply to other enums, such as `CodingKey` conformances or `ResponseHeaders`.

liamnichols reviewed (commented):
Wow great work! Thanks for this! 

> Currently does not do any of the redundant protocol stripping that `entities.protocols` does.

I think that is fine. This is a good start 🙏 

--- 

Looks like there are some issues with the CI checks, but I'd love to get this merged once that is addressed 🙏

aaronsky committed:
Addressed feedback, regenerated snapshots

aaronsky commented:
@liamnichols I believe this both addresses the test failures and your feedback. The PR suddenly got huge because I needed to regenerate snapshots, since protocols are sorted by name now.

aaronsky commented:
@liamnichols any update on this?

aaronsky committed:
Enhanced default values

Added vscode to gitignore, bumped lockfile for tests

Ignore vscode configuration

wip: default values

wip: edgecases

wip: fixes

update snapshots

Improved reliability of default argument output

Update snapshots

Remove print statement

Handle the optional-single-variant case correctly

Add guards against adding default values when disabled

Add a test for default values being disabled

Added docs

Parameterize includeDefaultValues to Templates

reorganized codable derive bits. not sure if that's enough

update docs

Revert codable changes

Remove decoder-init generation for non-codingkey-optimized entities

try this for default value description instead

update edgecases tests

Revert new CharacterSet

Remove default value fallbacks from decoders per feedback

aaronsky committed:
Make AnyJSON Sendable

damien-nd commented:
Any update on this PR? I would like to add Sendable on enum like we can do on entities.protocols 🙏

nbapps commented:
Any update on this ?

mgrider commented:
I'll add my voice to the chorus here. It would be great to get this merged.

mgrider commented:
...although looking closer, I can see some reasons we wouldn't want this version of this feature. For one thing, why is `enums` expected to be a separate part of the config from `entities`? It should be parallel (at the same level as) `protocols` underneath `entities`, IMO.

Also, I'm not sure why this PR had to regenerate all those test results. That's probably something that should be figured out before this could get merged.
